### DISCUSSION TITLE
[Help]: How to use ? ValueID

### DISCUSSION THREAD

**Discussion by bakman2** [zwave-js/node-red-contrib-zwave-js#117] [Q&A]
The wiki describes:

_Get the list of ValueIDs for a Node_
```
let Message = {
    payload: {
        mode: "ValueAPI",
        node: 2,
        method: "getDefinedValueIDs",
    }
}
return Message
````

_getValueMetadata_
```
let Message = {
    payload: {
        mode: "ValueAPI",
        node: 2,
        method: "getValueMetadata",
        params: [<ValueID>]
    }
}
return Message
```

I send (preselected 1 specific node) :

```
{
        mode: "ValueAPI",
        method: "getDefinedValueIDs",
    }
```
I get 2 (duplicated) messages back with an array (?)

Lets say I need this part from the array:
```
{"commandClassName":"Thermostat Mode","commandClass":64,"endpoint":0,"property":"manufacturerData","propertyName":"manufacturerData"}
```
I want to get the `getValueMetadata`, so I send:
```
{
        mode: "ValueAPI",
        method: "getValueMetadata",
        params: [64]   
}
```
Result:
```
"ZWaveError: Invalid ValueID passed!"
```

What is the ValueID here, this is not clear to me.

### Any more details?

If this is specific to some hardware or specific software version, please explain here.

### Version

6.1.1

### Node-RED Version

2.0.6

### What hardware are you using?

Other Linux

**Comment by marcus-j-davies** [comment]
Hi @bakman2,

The ValueID must be passed as an object - not an index of the array.


So to get a value for a given ValueID.

```javascript
let MyValueID = {
  "commandClassName": "Thermostat Mode",
  "commandClass": 64,
  "endpoint": 0,
  "property": "manufacturerData",
  "propertyName": "manufacturerData"
}

let Message = {
    payload: {
        mode: "ValueAPI",
        node: 2,
        method: "getValue",
        params: [MyValueID]
    }
}
return Message
```

`getDefinedValueIDs`
gives you the available ValueID's - but you must use them in there entirety (and not based on the index they appear)

**Comment by bakman2** [comment]
Ok thanks that works, that is not what I was expecting (I have never seen an "ID" been described as an object). Note I assumed that the commandclass was the valueid.

So how would I update a "ValueID" ? You pass in the same full object and as value the same full object with updated values ?
I am sorry for the silly questions, but the wiki is not clear to me with examples.

**Comment by marcus-j-davies** [comment]
Yeah, the backend driver refers to an ID - so I do agree its a little confusing.

so, to update a value using the ValueAPI

```javascript
let MyValueID = {
   .......
}

let Message = {
    payload: {
        mode: "ValueAPI",
        node: 2,
        method: "setValue",
        params: [MyValueID, your-value]
    }
}
return Message
```

the CCAPI - is easier, if you want to start there, but by all means go with ValueAPI if you want.

**Comment by marcus-j-davies** [comment]
> I am sorry for the silly questions, but the wiki is not clear to me with examples.

don't be sorry 👍

**Comment by bakman2** [comment]
I want the easiest path, I went in this completely blind, I need to migrate from another zwave node-red node. 
So I wanted to "translate" my previous flow to this. 

Tbh it is easier to look in the cache files than to retrieve the values via node-red.

Now i will look into the "CCAPI". Thanks again for the quick support.

**Comment by marcus-j-davies** [comment]
CCAPI - will require a device to be awake to retrieve data at that moment (or it will be retuned to you, when the device wakes up)

where-as the ValueAPI will use stored values (at least for returning data)
the stored values are updated - when your device send updates of course.

but there is no need for the device to be awake to fetch data if using ValueAPI

CCAPI is easier to use - but it does have draw backs (for battery devices) - at least if you want to fetch data, due to the wait that may occur

**Comment by marcus-j-davies** [comment]
@bakman2 
I have moved this to a discussion  - so holla if you need anything 👍

**Comment by bakman2** [comment]
Thanks. 

Shouldn't one be able to build a single array of objects for the whole topology in the network with all their properties and values ? It boggles my mind why these specifics are split up in by metadata and values separately as they belong together. Then again I only have 2 zwave devices so don't mind my oversimplification. 

I tried:

getDefinedValueIDs > split > for each getValueMetadata, but this ends up in an infinite loop with errors that not all ValueIDs cannot be found somewhere in between the messages, but I digress.

I think it could be more user friendly to provide additional nodes that do the translation mapping in the background with dynamically populated fields, something like:

![Screenshot 2021-10-01 at 13 06 23](https://user-images.githubusercontent.com/333960/135610299-2ec7becd-2d1e-4742-90ad-c8983906f1f2.jpg)

where the commands are dynamically retrieved based on node selection
and the input payload input suggestion based on the command selection as a translation layer to generate the messages

It is all kind of convoluted in my eyes, especially when dealing with the valueAPI - i am using battery powered devices.

**Reply by marcus-j-davies** [reply]
> Thanks.
> 
> Shouldn't one be able to build a single array of objects for the whole topology in the network with all their properties and values ? It boggles my mind why these specifics are split up in by metadata and values separately as they belong together. Then again I only have 2 zwave devices so don't mind my oversimplification.
> 
> I tried:
> 
> getDefinedValueIDs > split > for each getValueMetadata, but this ends up in an infinite loop with errors that not all ValueIDs cannot be found somewhere in between the messages, but I digress.
> 
> I think it could be more user friendly to provide additional nodes that do the translation mapping in the background with dynamically populated fields, something like:
> 
> ![Screenshot 2021-10-01 at 13 06 23](https://user-images.githubusercontent.com/333960/135610299-2ec7becd-2d1e-4742-90ad-c8983906f1f2.jpg)
> 
> where the commands are dynamically retrieved based on node selection and the input payload input suggestion based on the command selection as a translation layer to generate the messages
> 
> It is all kind of convoluted in my eyes, especially when dealing with the valueAPI - i am using battery powered devices.

As mentioned I am working on this ability already here.
https://github.com/zwave-js/node-red-contrib-zwave-js/issues/113

**Comment by marcus-j-davies** [comment]
So,

A ValueID describes:
The CC, the Nodes endpoint and the property for that CC.

example.

You have a device that has 2 batteries.
Each battery is served by a different endpoint (think of a Z-wave switch with 2 outlets = 2 endpoints, but still one node)

Each battery has the % level, charging state and a isLow property.
So each battery has 3 values (each of these values Is given a valueID)

but we have 2 batteries, so we need 6 ValueIDs (and the only real difference is the endpoint)
a valueID also includes the endpoint.

HINT 
double click a value in the UI - you will see its ValueID 😄 

![image](https://user-images.githubusercontent.com/55892693/135614293-6d57d742-5fae-47d8-b7ac-1a0ae845af89.png)

**Comment by marcus-j-davies** [comment]
PS I have a node being developed where it will generate commands for you 👍 
https://github.com/zwave-js/node-red-contrib-zwave-js/issues/113

**Comment by bakman2** [comment]
Hold on one second, i was not even aware there was a tab!

**Reply by marcus-j-davies** [reply]
Really? - I also added a network map.

Happy Friday! 😆

**Comment by marcus-j-davies** [comment]
https://github.com/zwave-js/node-red-contrib-zwave-js/wiki/User-Interface

**Comment by marcus-j-davies** [comment]
My screenshot is showing my development version - new icons and such for node status'

![image](https://user-images.githubusercontent.com/55892693/135615305-2ee21d9c-bd76-4eed-87b6-502ebabbeaa2.png)

**Comment by bakman2** [comment]
It all looks great, fantastic stuff. If the "basic editing" could also be translated to separate nodes, or something like that, that would be awesome. Let me first try to get this all working, i am able to turn on the valves and set the setpoints (took me a while tbh).

**Reply by marcus-j-davies** [reply]
the new `cmd-factory` node will;

 - Ask for the CC (dropdown)
 - Ask for the method on that CC (dropdown)
 - and will take JSONata values to dictate the params for the method, node and endpoint if needed.
 
 the `cmd-factory` node will then spit the command out to its sibling - which you can send directly to the controller or utilise in another node from NR

**Reply by bakman2** [reply]
This looks indeed what i was referring to, really nice!

**Reply by marcus-j-davies** [reply]
> This looks indeed what i was referring to, really nice!

I am not far from doing this - if you want to be a beta tester when its ready (and your still with us 😉)
I have 2 beta testers - but more the merrier of course.

**Comment by marcus-j-davies** [comment]
Do look at the CCAPI - its much much much easier (to get started with)
The only draw back in CCAPI - is that the UI wont update itself (unless the device acknowledges it with an update event)

You can force a UI update from CCAPI using `forceUpdate` I want bombard you, but it's on the wiki.

changing the thermo 
https://zwave-js.github.io/node-zwave-js/#/api/CCs/ThermostatSetpoint?id=set

```javascript
let Message = {
    payload: {
        mode: "CCAPI",
        node: 56,
        cc: "Thermostat Setpoint",
        method: "set",
        enums: {
            0: "ThermostatSetpointType"
        },
        params: ["Heating", 15, 0x00] /*0x00 - for Celsius */
    }
}
return Message
```

**Reply by bakman2** [reply]
Ok yes this saves on the whole valueid thing (which i still find quite odd). I have to investigate still on the params, but it makes a little more sense.

**Reply by marcus-j-davies** [reply]
The ValueID thing is more for UI's - that's my take on it - the UI tab is using it.
I just exposed because - well why not 😆 

some users do prefer using it though.

**Reply by bakman2** [reply]
>` params: ["Heating", 15, 0x00] /*0x00 - for Celsius */`

Where does the "Heating" come from ?
The (my) thermo has 2 things that need to be updated, 1) general state ("Off" or "Heat") and 2) the temperature value. (2 individual messages currently)

**Reply by marcus-j-davies** [reply]
if your thermo also does cooling - you can set the associated setpoint.

https://github.com/zwave-js/node-zwave-js/blob/3a7251d111320190f8936b277a723c0b0ae9711f/packages/zwave-js/src/lib/commandclass/ThermostatSetpointCC.ts#L54

**Reply by marcus-j-davies** [reply]
As this CC uses enums, we say pass parameter 0 to enum `ThermostatSetpointType`

```javascript
let Message = {
    payload: {
        mode: "CCAPI",
        node: 56,
        cc: "Thermostat Setpoint",
        method: "set",
        enums: {
            0: "ThermostatSetpointType"
        },
        params: ["Heating", 15, 0x00] /*0x00 - for Celsius */
    }
}
return Message
```

**Reply by bakman2** [reply]
It dazzles me tbh, it all goes way over my head, it appears to be a (over?)complicated protocol that is lacking a simple translation layer. I have no idea what a "enum" is, but looking at the source you pointed at and the example, it seems that the "Heating" is redundant as you specified the enum ?

**Reply by marcus-j-davies** [reply]
the zwave protocol has integers for the type of setpoint (its not the only CC in the spec)

	"N/A" = 0x00,
	"Heating" = 0x01, // CC v1
	"Cooling" = 0x02, // CC v1
	"Furnace" = 0x07, // CC v1
	"Dry Air" = 0x08, // CC v1
	"Moist Air" = 0x09, // CC v1
	"Auto Changeover" = 0x0a, // CC v1
	"Energy Save Heating" = 0x0b, // CC v2
	"Energy Save Cooling" = 0x0c, // CC v2
	"Away Heating" = 0x0d, // CC v2
	"Away Cooling" = 0x0e, // CC v3
	"Full Power" = 0x0f, // CC v3
	
instead of remembering these numbers, we can just give it a name like "Heating"
but then we need to convert "Heating" into the proper zwave spec numerical number.

we do that with an enum
But what enum should we use... we say that here.

  ```javascript
enums: {
   0: "ThermostatSetpointType"
}
```

so pass parameter 0 (zero) though the enum of `ThermostatSetpointType`
what then gets sent to the device for the type of Set point is 0x01 ("Heating")
if we were to set parameter 0 to 'Dry Air' - 0x08 will be used in the message to the device

**Reply by marcus-j-davies** [reply]
a thermostat can control anything from central heating, cooling, a furnace etc etc hence the need to specify the setpoint type.
indeed the thermo may only do one i.e Heating. but the specification still insists it needs to be sent 🤷

**Reply by marcus-j-davies** [reply]
here is all the CC's that you can use (albeit if your have such devices)

https://zwave-js.github.io/node-zwave-js/#/api/CCs/index

**Reply by bakman2** [reply]
As a sidenote; with the previous node I could send the following, which clicked in my mind. 
```
{"nodeid":3,"cmdclass":67,"cmdidx":1,"value":23}
```

This node is more verbose in a complex manner, but I got it working ;) thanks for your time and effort for creating it!

**Reply by marcus-j-davies** [reply]
Well if you do stick around 😉 the `cmd-factory` node will offer a very slimline approach to it all
(have been asked of this by current users - so I better do it 😆 )

BTW, if using the CCAPI - you dont have to use the CC Name, you can use the hexadecimal ID, as displayed here if you prefer
https://zwave-js.github.io/node-zwave-js/#/api/CCs/index

```javascript
let Message = {
    payload: {
        mode: "CCAPI",
        cc: 0x43,
        method: "set",
        enums: {
            0: "ThermostatSetpointType"
        },
        params: ["Heating", 15, 0x00]
    }
}
return Message
```

believe it or not, ZWave-JS - the backend driver to this, is the only open-source driver that strictly follows specifications, and one that offers the complete package Including S2 security) - click on include/exclude in the UI.

https://github.com/zwave-js

(and I do agree, in that it strips away a lot of colour)

I think my aim here was to have a 1:1 relationship with the driver.
it takes a little leaning  - but it ends up being a very robust setup  

[endSalesPitch]

😉 

any problems - please reach out to us.

**Comment by marcus-j-davies** [comment]
@bakman2,

we will soon have a much friendlier way to produce commands.
this allows a drop down of your target CC and the method to execute on that cc.

![image](https://user-images.githubusercontent.com/55892693/136087063-2cfd2d19-3082-488a-bdf5-abfc0e085679.png)
